### PR TITLE
Implemented copyWithZone, mutableCopyWithZone methods for various collection classes

### PR DESCRIPTION
**Pull request by sgl0v**
Implemented `copyWithZone`, `mutableCopyWithZone` methods for the `NSArray`, `NSMutableArray`, `NSDictionary`, `NSMutableDictionary`, `NSSet`, `NSMutableSet` classes

### CODE REVIEW

File: Foundation/NSArray.swift [hunk 1]
```diff
@@ -84,7 +84,7 @@ public class NSArray : NSObject, NSCopying, NSMutableCopying, NSSecureCoding, NS
     }
     
     public func copyWithZone(zone: NSZone) -> AnyObject {
-        return self
+        return NSArray(array: self.allObjects)
```
Comment (phausler) [inline Foundation/NSArray.swift]: NSArray is immutable so it just should return self, subclasses need to do the O(n) copy

File: Foundation/NSDictionary.swift [hunk 2]
```diff
@@ -152,7 +152,7 @@ public class NSDictionary : NSObject, NSCopying, NSMutableCopying, NSSecureCodin
     }
     
     public func copyWithZone(zone: NSZone) -> AnyObject {
-        NSUnimplemented()
+        return NSDictionary(objects: self.allValues, forKeys: self.allKeys as! [NSObject])
```
Comment (phausler) [inline Foundation/NSDictionary.swift]: this should just return self, same thing as NSArray; NSMutableDictionary and other subclasses should use the O(n) copy and NSDictionary (immutable) should just be a O(1) return of self

**Comment by sgl0v** [comment]
Updated PR according to the comments above.

**Comment by phausler** [comment]
Unfortunately we have a compiler crash in this one:
```
1.	While emitting SIL for 'copyWithZone' at Foundation/NSDictionary.swift:154:12
```

**Comment by sgl0v** [comment]
Fixed the compiler crash by replacing casting `self.allKeys as! [NSObject]` with `self.allKeys.map({ $0 as! NSObject})`

**Comment by phausler** [comment]
looks like that got it! that one keeps cropping up sadly

**Pull request merged**